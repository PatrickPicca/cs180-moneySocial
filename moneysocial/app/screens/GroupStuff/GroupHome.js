import React, { useState, useEffect } from 'react';
import { SafeAreaView, StyleSheet, Text, View, Dimensions, TextInput, Pressable, TouchableOpacity, Alert } from "react-native";
//import Svg, { Image, Ellipse, ClipPath } from "react-native-svg";
//import Animated, { useSharedValue, useAnimatedStyle, interpolate, withTiming, withDelay } from 'react-native-reanimated';
//import Ionicons from 'react-native-vector-icons/Ionicons';
//import CreateExpenseScreen from '../createExpenseScreen';
import colors from '../../config/colors';
//import { useNavigation, useFocusEffect } from '@react-navigation/native';
//import { API, graphqlOperation, Auth } from "aws-amplify";
//import { getGroup } from '../../../src/graphql/queries';
//import * as queries from '../../../src/graphql/queries';
//import * as mutations from '../../../src/graphql/mutations';
import awsconfig from '../../../src/aws-exports';
//import { AuthPiece } from 'aws-amplify-react-native';

function GroupHome(props) {

  const { id } = props;
  //const navigation = useNavigation();
  const [group, setGroup] = useState(null);

  /*useFocusEffect(
    React.useCallback(() => {
      fetchGroup();
    }, [])
  );

  const fetchGroup = async () => {
    try {
      const result = await API.graphql(graphqlOperation(getGroup, { id: id }));
      setGroup(result?.data?.getGroup);
    } catch (error) {
      console.log('Error fetching user groups', error);
    }
  };

  const handleCreateGroupExpenseScreen = (id) => {
    navigation.navigate('CreateGroupExpense', {id: id});
  }

  //deleteGroupHandler is not completed here because
  //not everyone should be able to delete the group, yet we lack
  //a groupOwner field, so I just merged it with leaveGroupHandler.
  //When the last person leaves the group, the group is automatically deleted.

  // const deleteGroupHandler = async (id) => {
  //   console.log("In delete group handler");
  //   Alert.alert(
  //     "Delete Group",
  //     "Are you sure you want to delete this group?",
  //     [
  //       {
  //         text: "Cancel",
  //         onPress: () => console.log("Cancel Pressed"),
  //         style: "cancel"
  //       },
  //       { text: "OK", onPress: () => {
  //         console.log("OK Pressed");
  //         let variables = {
  //           groupId: id
  //         }
  //         let result = API.graphql({query: queries.userGroupsByGroupId, variables: variables});
  //         let users = result.data.userGroupsByGroupId.items;
          


  const leaveGroupHandler = async (id) => {
        console.log("In delete group handler");
        let variables = {
          groupId: id
        }
        let result = await API.graphql({query: queries.userGroupsByGroupId, variables: variables});
        let recordId;
        // console.log(result.data.userGroupsByGroupId.items.length);
        let authUser = await Auth.currentAuthenticatedUser({bypassCache: true});
        let userInfo = await API.graphql({query: queries.getUser, variables: {id: authUser.attributes.sub}});
        let userId = userInfo.data.getUser.id;
        let found = false;
        for(let i = 0; i < result.data.userGroupsByGroupId.items.length; i++) {
          if(result.data.userGroupsByGroupId.items[i].userId == userId) {
            recordId = result.data.userGroupsByGroupId.items[i].id;
            found = true;
            break;
          }
        }
        if(!found) {
          Alert.alert("Error", "You are not a member of this group");
          return;
        }
        // console.log(userInfo.data.getUser.id);
        
        Alert.alert("Are you sure you want to leave the group \"" + group.name + "\"?", "You will no longer be able to see the group's expenses or add expenses to the group.", [
          {
            text: "Cancel",
            onPress: () => {
              console.log("Cancel Pressed");
              return;
            }
          }
          ,
          {
            text: "Ok",
            onPress: async () => {
            console.log("Ok Pressed");
            variables = {
              input: {
                id: recordId
              }
            }
            await API.graphql({query: mutations.deleteUserGroup, variables: variables});
            console.log("Successfully left group");
            if(result.data.userGroupsByGroupId.items.length == 1) {
              variables = {
                input: {
                  id: id
                }
              }
              await API.graphql({query: mutations.deleteGroup, variables: variables});
              console.log("Successfully deleted group");
            }
            navigation.goBack();
          }
        }
        ]);
        //if the user is the only member of the group, delete the group
  }

*/

  return (
    <SafeAreaView style={styles.container}>
      <Text style={styles.welcomeText} >{'Welcome ' + group?.name + '!'}</Text>

      <View style={styles.bottomContainer}>
        <Pressable style={styles.button} onPress={() => leaveGroupHandler(id, group)}>
          <Text style={styles.text}>{"Leave Group"}</Text>
        </Pressable>
      </View>
      
      <TouchableOpacity onPress={() => handleCreateGroupExpenseScreen(id)} style={styles.bottombutton}>
        <Text>+</Text>
      </TouchableOpacity>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-end',
    backgroundColor: colors.background,
  },
  welcomeText: {
    fontSize: 30,
    fontWeight: '600',
    color: 'black',
    letterSpacing: 0.5,
    paddingTop: 20,
    paddingLeft: 15,
    paddingBottom: 20
  },
  bottombutton: {
    backgroundColor: colors.primary,
    height: 55,
    width: 60,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 35,
    marginHorizontal: 5,
    marginVertical: 10,
    borderWidth: 2,
    borderColor: 'pink'
  },
  button: {
    backgroundColor: colors.accent,
    height: 40,
    width: 150,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius:5,
    marginHorizontal: 5,
    marginVertical: 10,
    borderWidth: 2,
    borderColor: 'white'
  },
})

export default GroupHome;
