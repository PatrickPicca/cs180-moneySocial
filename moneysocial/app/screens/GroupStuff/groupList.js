import React, { useState, useEffect, useContext } from 'react';
import { Animated, FlatList, StyleSheet, Text, Touchable, TouchableOpacity, View, ActivityIndicator } from 'react-native';
/*
import { API, graphqlOperation, Auth } from 'aws-amplify';
import { createGroup, createUserGroup } from '../../../src/graphql/mutations';
import { getGroup, listUserGroups, userGroupsByUserId } from '../../../src/graphql/queries';
*/
import { useNavigation, useFocusEffect } from '@react-navigation/native';
import colors from '../../config/colors';
import { SafeAreaView } from 'react-native-safe-area-context';

export default function GroupList({testLoad, testGroups}) {
  
  //const navigation = useNavigation();

  const handleGroupScreen = () => {
    //navigation.navigate("CreateGroupScreen");
    console.log('navigated');
  }
  const handleJoinScreen = () => {
    //navigation.navigate("JoinGroupScreen");
    console.log('navigated');

  }

  const [user, setUser] = useState(null);
  const [groups, setGroups] = useState(testGroups);
  const [userGroups, setUserGroups] = useState([]);
  const [loading, setLoading] = useState(testLoad);
  
  /*
  useFocusEffect(
    React.useCallback(() => {
      fetchUserGroups();
    }, [])
  );
  */
 /*
    const fetchUserGroups = async () => {

      try {
        
        const userInfo = await Auth.currentAuthenticatedUser();
        const userId = userInfo.attributes.sub;
        const result = await API.graphql(graphqlOperation(userGroupsByUserId, { userId }));
        const userGroupsData = result.data.userGroupsByUserId.items;
        setUserGroups(userGroupsData);
        const groupIds = userGroupsData.map(group => group.groupId);
        console.log(groupIds);
        const groups = [];
        for (const groupId of groupIds) {
          const group = await API.graphql(graphqlOperation(getGroup, { id: groupId }));
          groups.push(group.data.getGroup);
        }
        setGroups(groups);
        console.log(groups);
        setLoading(false);
      } catch (error) {
        console.log('Error fetching user groups', error);
      }
    };
*/
    const handleGroupDisplay = (id) => {
      //navigation.navigate("GroupTabs", { id: id });
      console.log('navigated');
    }
    const renderGroup = ({ item }) => {
      return (
        <TouchableOpacity onPress={() => handleGroupDisplay(item.id)} style={styles.groupItem}>
          <Text style={styles.groupName}>{item.name}</Text>
          <Text style={styles.keyText}>Key: {item.groupKey}</Text>
          <Text style={styles.idText}>Group ID: {item.id}</Text>
        </TouchableOpacity>
      );
    };
    
  
    if (loading) {
      return <Text>Loading...</Text>;
    }
  
    return (
      <SafeAreaView style={styles.container}>
          <FlatList
          data={groups}
          renderItem={renderGroup}
          keyExtractor={item => item.id}
          contentContainerStyle={{ flexGrow: 1 }}
          ListEmptyComponent={<Text>No user groups found</Text>}
          style={styles.groupListContainer}
          />

        <View style={styles.buttonContainer}>
          <TouchableOpacity style={styles.joinButton} onPress={handleJoinScreen}>
            <Text style={styles.createText}>Join Group</Text>
          </TouchableOpacity>
          <TouchableOpacity style={styles.createButton} onPress={handleGroupScreen}>
            <Text style={styles.createText}>Create Group</Text>
          </TouchableOpacity>
        </View>
        
      </SafeAreaView>
      
    );
  };


const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.background,
  },
  groupListContainer: {
    paddingVertical: 15,
    paddingHorizontal: 8,
  },
  groupItem: {
    justifyContent: 'space-between',
    marginVertical: 8,
    padding: 10,
    backgroundColor: colors.primary,
    borderRadius: 10,
  },
  groupName: {
    fontWeight: 'bold',
    fontSize: 16,
    color: '#fff',
  },
  keyText: {
    color:'#fff',
    fontSize: 14,
  },
  idText: {
    fontSize: 10,
    color: '#fff',
  },
  createText: {
    fontWeight: 'bold',
    fontSize: 16,
    color: '#fff',
  },
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    padding: 5,
  },
  createButton: {
    backgroundColor: colors.primary,
    borderRadius: 10,
    padding: 10,
    alignItems: 'center',
    marginBottom: 10,
    width: '49%',
  },
  joinButton: {
    backgroundColor: colors.primary,
    borderRadius: 10,
    padding: 10,
    alignItems: 'center',
    marginBottom: 10,
    width: '49%',
  },
});