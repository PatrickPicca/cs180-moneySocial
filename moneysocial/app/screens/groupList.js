import React, {useState} from 'react';
import { StyleSheet, Text, View, Dimensions, TextInput, Pressable, FlatList, ScrollView } from "react-native";
import Svg, {Image, Ellipse, ClipPath} from "react-native-svg";
import Animated, {useSharedValue, useAnimatedStyle, interpolate, withTiming, withDelay} from 'react-native-reanimated';

import colors from '../config/colors'



const {width, height} = Dimensions.get('window');

function PersonalExpenseScreen(props) {
    const {height, width} = Dimensions.get('window');
    const imagePosition = useSharedValue(1);
    const [isRegistering, setIsRegistering] = useState(false);
    
    const myValue1 = 360;
    const myValue2 = 1050;


    
    //Here we will list all groups you are a part of, alongside any current, total expenses you have made for that group.
      //In the view section you should be able to view / do the following
          //A list of all groups you are in and the current total expenses for them.
            //A button option to enter that group to view expenses for the past option.
            //A button option to expenses by a certain category for the past month.
            //A button option to make an expense.
          //Button option to view total expense for the past month by category.
          //Button option to view the list of all individual expenses of all time.
            //This should take you to some Expense screen where you can fill out a varying list of details for that one expense.
          //Button option to exit the current group.
          //Display somewhere, probably the header, a code unique to that group that others need to join.

    
    //Need a button that exits the groupExpense screen, taking you to you to the PersonalExpense screen.
    //Need some join group prompt with textbox where a user can type in a unique group code.
      //On success that group gets added to the user's list of active groups.
      //On failure, no group is joined.


    const imageAnimatedStyle = useAnimatedStyle(() => {
        const interpolation = interpolate(imagePosition.value, [0,1], [-height/2, 0])
        return {
            transform: [{translateY: withTiming(interpolation, {duration: 1000})}]
        };
    });

    const buttonsAnimatedStyle = useAnimatedStyle (() =>{
        const interpolation = interpolate(imagePosition.value, [0,1], [250,0])
        return {
            opacity: withTiming(imagePosition.value, 500),
            transform: [{translateY: withTiming(interpolation, {duration: 1000})}]
        }
    })

    const closeButtonContainerStyle = useAnimatedStyle (() =>{
        const interpolation = interpolate(imagePosition.value, [0,1], [180,360])
        return {
            opacity: withTiming(imagePosition.value === 1 ? 0 : 1, {duration: 800}),
            transform: [{rotate: withTiming(interpolation + "deg", {duration: 1000})}]
        }
    })

    const formAnimatedStyle = useAnimatedStyle(() => {
        const interpolation = interpolate(imagePosition.value, [0,1], [-height/2 + 35,0])
        return {
            opacity: imagePosition.value === 0 ? withDelay(400, withTiming(1,{duration: 1000})) : withTiming(0, {duration: 300}),
            transform: [{translateY: withTiming(interpolation, {duration: 1000})}]
        }
    })

    const WelcomeScreenHandler = () => {
    //  props.navigation.navigate('WelcomeScreen');
    }
    const groupListHandler = () => {
      console.log("logging in");
      }

    const ExpenseListScreenHander = () => {
      props.navigation.navigate('ExpenseListScreen');
    }
    const HomeScreenHandler = () => {
      props.navigation.navigate('HomeScreen');
    }




    const DATA = [
      {
        id: 'bd7acbea-c1b1-46c2-aed5-3ad53abb28ba',
        title: 'First Item',
      },
      {
        id: '3ac68afc-c605-48d3-a4f8-fbd91aa97f63',
        title: 'Second Item',
      },
      {
        id: '58694a0f-3da1-471f-bd96-145571e29d72',
        title: 'Third Item',
      },
      {
        id: '5869d72',
        title: '6th Item',
      },
    ];

    const Item = ({title}) => (
      <View style={styles.item}>
        <Text style={styles.title}>{title}</Text>
        <Pressable style={styles.bottombutton} onPress={groupListHandler}>
          <Text style={styles.bottombuttonText}>Logout</Text>
        </Pressable>

      </View>
    );

    //View should have a list of all groups the user is in, as well as that groups current total expense.
    return (
      
      <View style={{flex:1}}>onPress
        <FlatList
          data={DATA}
          renderItem={({item}) => <Item title={item.title} />}
          keyExtractor={item => item.id}
        />
        <View style={styles.container}>

          <Animated.View style={buttonsAnimatedStyle}>
            <View style={styles.bottomScreenHeader}>
              <Pressable style={styles.bottombutton} onPress={WelcomeScreenHandler}>
                <Text style={styles.bottombuttonText}>Logout</Text>
              </Pressable>

              <Pressable style={styles.bottombutton} onPress={HomeScreenHandler}>
                <Text style={styles.bottombuttonText}>Home</Text>
              </Pressable>

              <Pressable style={styles.bottombutton} onPress={ExpenseListScreenHander}>
                <Text style={styles.bottombuttonText}>Details</Text>
              </Pressable>

              
              
            </View>  
          </Animated.View>
        </View>
      </View>
    );
    
}

/* 
          <ScrollView style={styles.scrollView}>
            <View style={styles.topContainer}>

              <View style={styles.displayBalance}>
                <Text style={styles.displayText}>{'Group 1: $' + myValue1}</Text>
              </View>
              <View style={styles.displayBalance}>
                <Text style={styles.displayText}>{'Group 2: $' + myValue2}</Text>
              </View>
            </View>
          </ScrollView>
          <Animated.View style={[StyleSheet.absoluteFill, imageAnimatedStyle]}>

          </Animated.View> 
*/

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'flex-end',
        backgroundColor: colors.background,
      },
      button: {
        
        backgroundColor: colors.primary,
        height: 55,
        width: 60,
        alignItems: 'center',
        justifyContent: 'center',
        borderRadius: 35,
        marginHorizontal: 5,
        marginVertical: 10,
        borderWidth: 2,
        borderColor: 'white'
      },
      bottombutton: {
        
        backgroundColor: colors.primary,
        height: 55,
        width: 60,
        alignItems: 'center',
        justifyContent: 'center',
        borderRadius: 35,
        marginHorizontal: 5,
        marginVertical: 10,
        borderWidth: 2,
        borderColor: 'pink'
      },
      buttonText: {
        fontSize: 20,
        fontWeight: '600',
        color: 'white',
        letterSpacing: 0.5
      },
      bottombuttonText: {
        fontSize: 15,
        fontWeight: '600',
        color: 'white',
        letterSpacing: 0.5
      },
      bottomContainer: {
        justifyContent: 'center',
        height: height / 2,
      },
      bottomScreenHeader: {
        flexDirection: 'row',
        justifyContent: 'center',
        height: height / 8,
      },
      textInput: {
        height: 50,
        borderWidth: 1,
        borderColor: colors.primary,
        marginHorizontal: 20,
        marginVertical: 10,
        borderRadius: 25,
        paddingLeft: 10
      },
      formButton: {
        backgroundColor: colors.primary,
        height: 55,
        alignItems: 'center',
        justifyContent: 'center',
        borderRadius: 35,
        marginHorizontal: 20,
        marginVertical: 10,
        borderWidth: 1,
        borderColor: 'white',
        shadowColor: "#000",
        shadowOffset: {
          width: 0,
          height: 4,
        },
        shadowOpacity: 0.25,
        shadowRadius: 3.84,
        elevation: 5,
      },
      topContainer: {
        justifyContent: 'center',
        height: height / 4,
        top: 50,
      },
      formInputContainer: {
        marginBottom: 70,
        ...StyleSheet.absoluteFill,
        zIndex: -1,
        justifyContent: 'center'
      },
      closeButtonContainer: {
        height: 40,
        width: 40,
        justifyContent: 'center',
        alignSelf: 'center',
        shadowColor: "#000",
        shadowOffset: {
          width: 0,
          height: 5,
        },
        shadowOpacity: 0.34,
        shadowRadius: 6.27,
        elevation: 1,
        backgroundColor: 'white',
        alignItems: 'center',
        borderRadius: 20,
        top: -25
    }
})

export default PersonalExpenseScreen;