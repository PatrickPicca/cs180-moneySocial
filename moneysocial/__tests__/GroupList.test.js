import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import GroupList from '../app/screens/GroupStuff/groupList';
import { TouchableOpacity, Text } from 'react-native';
import renderer from 'react-test-renderer';
import { renderGroup } from '../app/screens/GroupStuff/groupList';


describe('GroupList component', () => {
  it('should render loading indicator if loading is true', () => {
    const { getByText } = render(<GroupList testLoad={true} />);
    expect(getByText('Loading...')).toBeDefined();
  });

  test('renders correctly', () => {
    const item = { id: 1, name: 'Group 1', groupKey: 'ABC' };
    const tree = renderer.create(<renderGroup item={item} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  
});
