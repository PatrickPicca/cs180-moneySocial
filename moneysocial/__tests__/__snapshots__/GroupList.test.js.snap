// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupList component renders correctly 1`] = `
<renderGroup
  item={
    Object {
      "groupKey": "ABC",
      "id": 1,
      "name": "Group 1",
    }
  }
/>
`;
